lib/common/format.h

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_FORMAT_H_
+#define FORTRAN_COMMON_FORMAT_H_
+
+#include "Fortran.h"
+#include <cstring>
+
+namespace Fortran::common {
+
+struct FormatError {
+  static constexpr int maxArgLength_{25};
+
+  const char *text_;  // error text; may have one %s argument
+  char arg_[maxArgLength_ + 1];  // optional %s argument value
+  int offset_;  // offset to error marker
+  int length_;  // length of error marker
+};
+
+template<typename CHAR = char> class FormatValidator {
+public:
+  explicit FormatValidator(const CHAR *format, size_t length, IoStmtKind stmt,
+      bool warnOnNonstandardUsage, FormatError *errors, int maxErrors)
+    : format_{format}, end_{format ? format + length : nullptr}, stmt_{stmt},
+      warnOnNonstandardUsage_{warnOnNonstandardUsage}, errors_{errors},
+      maxErrors_{maxErrors}, cursor_{format - 1} {}
+
+  int Check();
+
+private:
+  enum class TokenKind {
+    None,
+    A,
+    B,
+    BN,
+    BZ,
+    D,
+    DC,
+    DP,
+    DT,
+    E,
+    EN,
+    ES,
+    EX,
+    F,
+    G,
+    I,
+    L,
+    O,
+    P,
+    RC,
+    RD,
+    RN,
+    RP,
+    RU,
+    RZ,
+    S,
+    SP,
+    SS,
+    T,
+    TL,
+    TR,
+    X,
+    Z,
+    Colon,
+    Slash,
+    Backslash,  // nonstandard: inhibit newline on output
+    Dollar,  // nonstandard: inhibit newline on output on terminals
+    Star,
+    LParen,
+    RParen,
+    Comma,
+    Point,
+    Sign,
+    UnsignedInteger,  // value in integerValue_
+    String,  // char-literal-constant or hollerith constant
+  };
+
+  struct Token {
+    explicit Token() : kind_{TokenKind::None}, offset_{0}, length_{1} {}
```

Comment:
I'd put those initial values on the declarations of the data members themselves, and then remove the explicit constructor.

Reply:
Ok; done.